File: Alkitab/src/yuku/alkitab/base/ac/SecretSettingsActivity.java
package yuku.alkitab.base.ac;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.preference.Preference;
import yuku.alkitab.debug.R;
import yuku.alkitab.base.S;
import yuku.alkitab.base.ac.base.BasePreferenceActivity;
import yuku.alkitab.base.model.ProgressMark;
import yuku.alkitab.base.model.ProgressMarkHistory;
import yuku.alkitab.base.util.Sqlitil;

import java.util.List;

public class SecretSettingsActivity extends BasePreferenceActivity {
	public static final String TAG = SecretSettingsActivity.class.getSimpleName();

	private Preference.OnPreferenceClickListener secret_progress_mark_history_click = new Preference.OnPreferenceClickListener() {
		@Override
		public boolean onPreferenceClick(final Preference preference) {
			final List<ProgressMark> progressMarks = S.getDb().listAllProgressMarks();

			final String[] items = new String[progressMarks.size()];
			for (int i = 0; i < progressMarks.size(); i++) {
				final ProgressMark progressMark = progressMarks.get(i);
				items[i] = progressMark.caption + " (preset_id " + progressMark.preset_id + ")";
			}

			new AlertDialog.Builder(SecretSettingsActivity.this)
			.setItems(items, new DialogInterface.OnClickListener() {
				@Override
				public void onClick(final DialogInterface dialog, final int which) {
					final List<ProgressMarkHistory> pmhs = S.getDb().listProgressMarkHistoryByPresetId(progressMarks.get(which).preset_id);
					final String[] items = new String[pmhs.size()];
					for (int i = 0; i < pmhs.size(); i++) {
						final ProgressMarkHistory pmh = pmhs.get(i);
						items[i] = "'" + pmh.progress_mark_caption + "' " + Sqlitil.toLocaleDateMedium(pmh.createTime) + ": " + S.activeVersion.reference(pmh.ari);
					}
					new AlertDialog.Builder(SecretSettingsActivity.this)
					.setItems(items, null)
					.show();
				}
			})
			.show();

			return true;
		}
	};

	@SuppressWarnings("deprecation") @Override protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		
		addPreferencesFromResource(R.xml.secret_settings);

		findPreference("secret_progress_mark_history").setOnPreferenceClickListener(secret_progress_mark_history_click);
	}
}

File: Alkitab/src/main/java/yuku/alkitab/base/ac/SecretSettingsActivity.kt
package yuku.alkitab.base.ac

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import androidx.appcompat.widget.Toolbar
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import java.util.Locale
import yuku.alkitab.base.App
import yuku.alkitab.base.ac.base.BaseActivity
import yuku.alkitab.base.settings.SettingsActivity.Companion.autoDisplayListPreference
import yuku.alkitab.base.util.Sqlitil.toLocaleDateMedium
import yuku.alkitab.base.widget.ConfigurationWrapper
import yuku.alkitab.debug.R

class SecretSettingsActivity : BaseActivity() {
    class SecretSettingsFragment : PreferenceFragmentCompat() {
        private val secret_progress_mark_history_click = Preference.OnPreferenceClickListener {
            val progressMarks = App.services.storage.db.listAllProgressMarks()
            val labels = progressMarks.map { "${it.caption} (preset_id ${it.preset_id})" }.toTypedArray()
            MaterialAlertDialogBuilder(requireActivity())
                .setItems(labels) { _, index ->
                    val pmhs = App.services.storage.db.listProgressMarkHistoryByPresetId(progressMarks[index].preset_id)
                    val items = pmhs.map { "'${it.progress_mark_caption}' ${toLocaleDateMedium(it.createTime)}: ${App.services.versions.activeVersion().reference(it.ari)}" }.toTypedArray()
                    MaterialAlertDialogBuilder(requireActivity())
                        .setItems(items, null)
                        .show()
                }
                .show()
            true
        }

        private val secret_version_table_click = Preference.OnPreferenceClickListener {
            val items = App.services.storage.db.listAllVersions().map { mv ->
                String.format(
                    Locale.US, "filename=%s preset_name=%s modifyTime=%s active=%s ordering=%s locale=%s shortName=%s longName=%s description=%s",
                    mv.filename, mv.preset_name, mv.modifyTime, mv.active, mv.ordering, mv.locale, mv.shortName, mv.longName, mv.description
                )
            }.toTypedArray()

            MaterialAlertDialogBuilder(requireActivity())
                .setItems(items, null)
                .show()
            true
        }

        private var secret_sync_debug = Preference.OnPreferenceClickListener {
            startActivity(Intent(App.context, SecretSyncDebugActivity::class.java))
            true
        }

        private val configurationPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, _ ->
            val handler = Handler(Looper.getMainLooper())

            // do this after this method returns true
            handler.post {
                ConfigurationWrapper.notifyConfigurationNeedsUpdate()

                // restart this activity
                activity?.recreate()
            }
            true
        }

        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            addPreferencesFromResource(R.xml.secret_settings)
            findPreference<Preference>("secret_progress_mark_history")?.onPreferenceClickListener = secret_progress_mark_history_click
            findPreference<Preference>("secret_version_table")?.onPreferenceClickListener = secret_version_table_click
            findPreference<Preference>("secret_sync_debug")?.onPreferenceClickListener = secret_sync_debug
            val pref_forceFontScale = findPreference<Preference>(getString(R.string.pref_forceFontScale_key)) as ListPreference?
            if (pref_forceFontScale != null) {
                pref_forceFontScale.onPreferenceChangeListener = configurationPreferenceChangeListener
                autoDisplayListPreference(pref_forceFontScale)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_secret_settings)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        setupEdgeToEdgeDisplay(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    companion object {
        fun createIntent(): Intent {
            return Intent(App.context, SecretSettingsActivity::class.java)
        }
    }
}
